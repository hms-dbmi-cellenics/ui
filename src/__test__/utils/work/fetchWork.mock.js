const mockGenesListData = ['A', 'B', 'C', 'D'];

// const mockGeneExpressionData = {
//   A: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//   },
//   B: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.702857143,
//       stdev: 2.551115536,
//       expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.702857143,
//       stdev: 2.551115536,
//       expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
//     },
//   },
//   C: {
//     rawExpression: {
//       min: 0,
//       max: 3.4,
//       mean: 1.68,
//       stdev: 2.141525936,
//       expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 3.4,
//       mean: 1.68,
//       stdev: 2.141525936,
//       expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
//     },
//   },
//   D: {
//     rawExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//     truncatedExpression: {
//       min: 0,
//       max: 6.8,
//       mean: 1.68,
//       stdev: 2.597331964,
//       expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
//     },
//   },
//   E: { hello: 'world' },
// };

// const mockCacheGet = jest.fn((x) => {
//   const mockCacheContents = {
//     fd3161a878f67ebf54018720cffd6a66: 'A', // pragma: allowlist secret
//     '10250a11679234110a1c260d6fd81d3c': 'B', // pragma: allowlist secret
//     f5c957411a28de68f35e1f5c8a29da7e: 'C', // pragma: allowlist secret
//     '33faa711a94a2028b5bae1778126aec0': 'E', // pragma: allowlist secret
//   };

//   if (x in mockCacheContents) {
//     return mockGeneExpressionData[mockCacheContents[x]];
//   }

//   return null;
// });

const mockCacheGet = jest.fn();
const mockCacheSet = jest.fn();
const mockCacheRemove = jest.fn();

const mockFetchWorkRequest = jest.fn(() => true);

const mockSeekFromS3 = jest.fn();

const mockCacheModule = {
  get: jest.fn((x) => mockCacheGet(x)),
  set: jest.fn((key, val) => mockCacheSet(key, val)),
  _remove: jest.fn((key) => mockCacheRemove(key)),
};

const mockFetchWorkModule = {
  __esModule: true,
  fetchWork: mockFetchWorkRequest,
};

const mockQcPipelineStartDate = '2021-01-01T01:01:01.000Z';

export {
  mockGenesListData, mockCacheGet, mockCacheSet,
  mockFetchWorkRequest, mockSeekFromS3,
  mockQcPipelineStartDate, mockCacheModule, mockFetchWorkModule,
};
