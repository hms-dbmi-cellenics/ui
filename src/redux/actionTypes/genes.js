const GENES = 'genes';

/**
 * Turns on the loading condition for gene information.
 * Components that have the particular properties requested in their dependencies
 * should set a loading state at this event.
 */
const GENES_PROPERTIES_LOADING = `${GENES}/propertiesLoading`;

/**
 * Sets the state of the gene data to be successfully loaded, with the appropriate gene data.
 * This is meant to be used for paginated data, where the data is
 * lazy-loaded by an API call or equivalent.
 */
const GENES_PROPERTIES_LOADED_PAGINATED = `${GENES}/propertiesLoadedPaginated`;

/**
 * Sets an error condition for gene properties data.
 */
const GENES_PROPERTIES_ERROR = `${GENES}/propertiesError`;

/**
 * Updates the list of selected genes with a new selected gene.
 */
const GENES_SELECT = `${GENES}/select`;

/**
 * Removes a deselected gene from the list of selected genes.
 */
const GENES_DESELECT = `${GENES}/deselect`;

/**
 * Turns on the loading state for expression data.
 */
const GENES_EXPRESSION_LOADING = `${GENES}/expressionLoading`;

/**
 * Sets the state of the expression store to be successfully loaded,
 * with the appropriate expression data.
 */
const GENES_EXPRESSION_LOADED = `${GENES}/expressionLoaded`;

/**
 * Sets an error condition for gene expression data.
 */
const GENES_EXPRESSION_ERROR = `${GENES}/expressionError`;

/**
 * Sets marker genes as loading
 */
const MARKER_GENES_LOADING = `${GENES}/markerGenesLoading`;

/**
 * Marker genes have finished loading
 */
const MARKER_GENES_LOADED = `${GENES}/markerGenesLoaded`;

/**
 * Marker genes load task had some kind of error
 */
const MARKER_GENES_ERROR = `${GENES}/markerGenesError`;

/**
 * Downsampled genes expression load task
 */
const DOWNSAMPLED_GENES_EXPRESSION_LOADING = `${GENES}/downsampledGenesExpressionLoading`;

/**
 * Downsampled genes expression loaded
 */
const DOWNSAMPLED_GENES_EXPRESSION_LOADED = `${GENES}/downsampledGenesExpressionLoaded`;

/**
 * Downsampled genes expression loaded
 */
const DOWNSAMPLED_GENES_EXPRESSION_ERROR = `${GENES}/downsampledGenesExpressionError`;

export {
  GENES_PROPERTIES_LOADING, GENES_PROPERTIES_LOADED_PAGINATED, GENES_PROPERTIES_ERROR,
  GENES_SELECT, GENES_DESELECT,
  MARKER_GENES_LOADING, MARKER_GENES_LOADED, MARKER_GENES_ERROR,
  GENES_EXPRESSION_LOADING, GENES_EXPRESSION_LOADED, GENES_EXPRESSION_ERROR,
  DOWNSAMPLED_GENES_EXPRESSION_LOADING,
  DOWNSAMPLED_GENES_EXPRESSION_LOADED,
  DOWNSAMPLED_GENES_EXPRESSION_ERROR,
};
