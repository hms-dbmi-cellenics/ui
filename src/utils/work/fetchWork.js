import { Environment, isBrowser } from 'utils/deploymentInfo';

import cache from 'utils/cache';
import dispatchWorkRequest from 'utils/work/dispatchWorkRequest';
import downloadFromS3 from 'utils/work/downloadFromS3';
import waitForWorkRequest from 'utils/work/waitForWorkRequest';

// retrieveData will try to get the data for the given experimentId and ETag from
// the fastest source possible. It will try to get the data in order from:
// 1. Browser cache
// 2. S3 (cache) via signedURL
// 3. Worker
//   3.1. Via socket (small data)
//   3.2. Via S3 (large data) via signedURL
const retrieveData = async (experimentId,
  ETag,
  signedUrl,
  request,
  timeout,
  body,
  dispatch,
  useCache) => {
  // 1. Check if we have the ETag in the browser cache (no worker)
  const cachedData = useCache ? await cache.get(ETag) : null;
  if (cachedData) {
    return cachedData;
  }

  // 2. Check if data is cached in S3 so we can download from the signed URL (no worker)
  if (signedUrl) {
    return await downloadFromS3(body.name, signedUrl);
  }

  // 3. If we don't have signedURL, wait for the worker to send us the data via
  // - the data via socket
  // - the signedURL to download the data from S3
  const { signedUrl: workerSignedUrl, data } = await waitForWorkRequest(
    ETag,
    experimentId,
    request,
    timeout,
    dispatch,
  );

  // 3.1. The worker send the data via socket because it's small enough
  if (data) {
    return data;
  }
  // 3.2. The worker send a signedUrl to download the data
  return await downloadFromS3(body.name, workerSignedUrl);
};

const fetchWork = async (
  experimentId,
  body,
  getState,
  dispatch,
  optionals = {},
) => {
  const {
    extras = undefined,
    timeout = 180,
    broadcast = false,
    onETagGenerated = () => { },
  } = optionals;

  if (!isBrowser) {
    throw new Error('Disabling network interaction on server');
  }

  const { environment } = getState().networkResources;
  if (environment === Environment.DEVELOPMENT && !localStorage.getItem('disableCache')) {
    localStorage.setItem('disableCache', 'true');
  }

  // 1. Contact the API to get ETag and possible S3 signedURL
  const { ETag, signedUrl, request } = await dispatchWorkRequest(
    experimentId,
    body,
    timeout,
    {
      broadcast,
      ...extras,
    },
    dispatch,
  );

  onETagGenerated(ETag);

  const useCache = body.name !== 'GeneExpression';

  // 2. Try to get the data from the fastest source possible
  const data = await retrieveData(
    experimentId, ETag, signedUrl, request, timeout, body, dispatch, useCache,
  );

  // 3. Cache the data in the browser
  if (useCache) {
    await cache.set(ETag, data);
  }

  return data;
};

export default fetchWork;
