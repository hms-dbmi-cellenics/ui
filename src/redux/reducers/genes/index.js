import getInitialState from 'redux/reducers/genes/getInitialState';
import {
  GENES_PROPERTIES_LOADING, GENES_PROPERTIES_LOADED_PAGINATED, GENES_PROPERTIES_ERROR,
  GENES_SELECT, GENES_DESELECT,
  GENES_EXPRESSION_LOADING, GENES_EXPRESSION_LOADED, GENES_EXPRESSION_ERROR,
  MARKER_GENES_LOADING, MARKER_GENES_LOADED, MARKER_GENES_ERROR,
  DOWNSAMPLED_GENES_EXPRESSION_LOADING,
  DOWNSAMPLED_GENES_EXPRESSION_LOADED,
  DOWNSAMPLED_GENES_EXPRESSION_ERROR,
} from 'redux/actionTypes/genes';

import { EXPERIMENT_SETTINGS_QC_START } from 'redux/actionTypes/experimentSettings';

import genesPropertiesLoading from 'redux/reducers/genes/genesPropertiesLoading';
import genesPropertiesLoadedPaginated from 'redux/reducers/genes/genesPropertiesLoadedPaginated';
import genesPropertiesError from 'redux/reducers/genes/genesPropertiesError';

import genesExpressionLoading from 'redux/reducers/genes/genesExpressionLoading';
import genesExpressionLoaded from 'redux/reducers/genes/genesExpressionLoaded';
import genesExpressionError from 'redux/reducers/genes/genesExpressionError';

import downsampledGenesLoading from 'redux/reducers/genes/downsampledGenesLoading';
import downsampledGenesLoaded from 'redux/reducers/genes/downsampledGenesLoaded';

import markerGenesLoading from 'redux/reducers/genes/markerGenesLoading';
import markerGenesError from 'redux/reducers/genes/markerGenesError';
import markerGenesLoaded from 'redux/reducers/genes/markerGenesLoaded';

import genesSelect from 'redux/reducers/genes/genesSelect';
import genesDeselect from 'redux/reducers/genes/genesDeselect';
import downsampledGenesError from './downsampledGenesError';

const genesReducer = (state = getInitialState(), action) => {
  switch (action.type) {
    case GENES_PROPERTIES_LOADING: {
      return genesPropertiesLoading(state, action);
    }
    case GENES_PROPERTIES_LOADED_PAGINATED: {
      return genesPropertiesLoadedPaginated(state, action);
    }
    case GENES_PROPERTIES_ERROR: {
      return genesPropertiesError(state, action);
    }
    case GENES_SELECT: {
      return genesSelect(state, action);
    }
    case GENES_DESELECT: {
      return genesDeselect(state, action);
    }
    case GENES_EXPRESSION_LOADING: {
      return genesExpressionLoading(state, action);
    }
    case GENES_EXPRESSION_LOADED: {
      return genesExpressionLoaded(state, action);
    }
    case GENES_EXPRESSION_ERROR: {
      return genesExpressionError(state, action);
    }
    case EXPERIMENT_SETTINGS_QC_START: {
      return getInitialState();
    }
    case MARKER_GENES_LOADING: {
      return markerGenesLoading(state, action);
    }
    case DOWNSAMPLED_GENES_EXPRESSION_LOADING: {
      return downsampledGenesLoading(state, action);
    }
    case DOWNSAMPLED_GENES_EXPRESSION_LOADED: {
      return downsampledGenesLoaded(state, action);
    }
    case DOWNSAMPLED_GENES_EXPRESSION_ERROR: {
      return downsampledGenesError(state, action);
    }
    case MARKER_GENES_LOADED: {
      return markerGenesLoaded(state, action);
    }
    case MARKER_GENES_ERROR: {
      return markerGenesError(state, action);
    }
    default: {
      return state;
    }
  }
};

export default genesReducer;
