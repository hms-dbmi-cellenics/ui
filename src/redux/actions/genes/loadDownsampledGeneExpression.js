import { SparseMatrix } from 'mathjs';

import {
  DOWNSAMPLED_GENES_EXPRESSION_LOADING,
  DOWNSAMPLED_GENES_EXPRESSION_LOADED,
  DOWNSAMPLED_GENES_EXPRESSION_ERROR,
} from 'redux/actionTypes/genes';

import fetchWork from 'utils/work/fetchWork';
import getTimeoutForWorkerTask from 'utils/getTimeoutForWorkerTask';

// Debounce so that we only fetch once the settings are done being set up
const loadDownsampledGeneExpression = (
  experimentId,
  genes,
  componentUuid,
  withHiddenCellSets = false,
) => async (dispatch, getState) => {
  const state = getState();

  const {
    groupedTracks,
    selectedCellSet,
    selectedPoints,
  } = state.componentConfig[componentUuid]?.config;

  const hiddenCellSets = withHiddenCellSets ? Array.from(state.cellSets.hidden) : [];

  const downsampleSettings = {
    selectedCellSet,
    groupedTracks,
    selectedPoints,
    hiddenCellSets,
  };

  if (genes.length === 0) {
    dispatch({
      type: DOWNSAMPLED_GENES_EXPRESSION_LOADED,
      payload: {
        componentUuid,
        genes,
      },
    });

    return;
  }

  const body = {
    name: 'GeneExpression',
    genes,
    downsampled: true,
    downsampleSettings,
  };

  const timeout = getTimeoutForWorkerTask(getState(), 'GeneExpression');

  try {
    let requestETag;

    const {
      orderedGeneNames,
      rawExpression: rawExpressionJson,
      truncatedExpression: truncatedExpressionJson,
      zScore: zScoreJson,
      stats,
      cellOrder,
    } = await fetchWork(
      experimentId, body, getState, dispatch,
      {
        timeout,
        onETagGenerated: (ETag) => {
          requestETag = ETag;

          // Dispatch loading state.
          dispatch({
            type: DOWNSAMPLED_GENES_EXPRESSION_LOADING,
            payload: {
              experimentId,
              componentUuid,
              genes,
              ETag,
            },
          });
        },
      },
    );

    // If the ETag is different, that means that a new request was sent in between
    // So we don't need to handle this outdated result
    if (getState().genes.expression.downsampled.ETag !== requestETag) {
      return;
    }

    const rawExpression = SparseMatrix.fromJSON(rawExpressionJson);
    const truncatedExpression = SparseMatrix.fromJSON(truncatedExpressionJson);
    const zScore = SparseMatrix.fromJSON(zScoreJson);

    dispatch({
      type: DOWNSAMPLED_GENES_EXPRESSION_LOADED,
      payload: {
        componentUuid,
        genes,
        newGenes: {
          orderedGeneNames,
          stats,
          rawExpression,
          truncatedExpression,
          zScore,
          cellOrder,
        },
      },
    });
  } catch (error) {
    dispatch({
      type: DOWNSAMPLED_GENES_EXPRESSION_ERROR,
      payload: {
        experimentId,
        componentUuid,
        genes,
        error,
      },
    });
  }
};

export default loadDownsampledGeneExpression;
