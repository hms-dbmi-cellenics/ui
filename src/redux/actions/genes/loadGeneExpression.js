import _ from 'lodash';
import { SparseMatrix } from 'mathjs';

import {
  GENES_EXPRESSION_LOADING, GENES_EXPRESSION_ERROR, GENES_EXPRESSION_LOADED,
} from 'redux/actionTypes/genes';

import fetchWork from 'utils/work/fetchWork';
import getTimeoutForWorkerTask from 'utils/getTimeoutForWorkerTask';
import upperCaseArray from 'utils/upperCaseArray';

const findLoadedGenes = (matrix, selectedGenes) => {
  // Check which of the genes we actually need to load. Only do this if
  // we are not forced to reload all of the data.
  const storedGenes = matrix.getStoredGenes();
  const genesToLoad = [...selectedGenes].filter(
    (gene) => !new Set(upperCaseArray(storedGenes)).has(gene.toUpperCase()),
  );

  const genesAlreadyLoaded = storedGenes.filter(
    (gene) => upperCaseArray(selectedGenes).includes(gene.toUpperCase()),
  );

  return { genesToLoad, genesAlreadyLoaded };
};

const loadGeneExpression = (
  experimentId,
  genes,
  componentUuid,
) => async (dispatch, getState) => {
  const { loading, matrix } = getState().genes.expression.full;

  // If other gene expression data is already being loaded, don't dispatch.
  if (loading.length > 0) {
    return null;
  }
  const { genesToLoad, genesAlreadyLoaded } = findLoadedGenes(matrix, genes);

  if (genesToLoad.length === 0) {
    // All genes are already loaded.
    return dispatch({
      type: GENES_EXPRESSION_LOADED,
      payload: {
        experimentId,
        componentUuid,
        genes: genesAlreadyLoaded,
      },
    });
  }

  // Dispatch loading state.
  dispatch({
    type: GENES_EXPRESSION_LOADING,
    payload: {
      experimentId,
      componentUuid,
      genes,
    },
  });

  const body = {
    name: 'GeneExpression',
    genes: genesToLoad,
    downsampled: false,
  };

  const timeout = getTimeoutForWorkerTask(getState(), 'GeneExpression');

  try {
    const {
      orderedGeneNames,
      rawExpression: rawExpressionJson,
      truncatedExpression: truncatedExpressionJson,
      zScore: zScoreJson,
      stats,
    } = await fetchWork(
      experimentId, body, getState, dispatch, { timeout },
    );

    const rawExpression = SparseMatrix.fromJSON(rawExpressionJson);
    const truncatedExpression = SparseMatrix.fromJSON(truncatedExpressionJson);
    const zScore = SparseMatrix.fromJSON(zScoreJson);

    const fetchedGenes = _.concat(genesAlreadyLoaded, orderedGeneNames);

    dispatch({
      type: GENES_EXPRESSION_LOADED,
      payload: {
        componentUuid,
        genes: fetchedGenes,
        newGenes: {
          orderedGeneNames,
          stats,
          rawExpression,
          truncatedExpression,
          zScore,
        },
      },
    });
  } catch (error) {
    dispatch({
      type: GENES_EXPRESSION_ERROR,
      payload: {
        experimentId,
        componentUuid,
        genes,
        error,
      },
    });
  }
};

export default loadGeneExpression;
