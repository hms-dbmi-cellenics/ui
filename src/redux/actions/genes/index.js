import loadGeneExpression from './loadGeneExpression';
import changeGeneSelection from './changeGeneSelection';
import loadPaginatedGeneProperties from './loadPaginatedGeneProperties';
import loadMarkerGenes from './loadMarkerGenes';
import loadDownsampledGeneExpression from './loadDownsampledGeneExpression';

export {
  loadGeneExpression,
  changeGeneSelection,
  loadPaginatedGeneProperties,
  loadMarkerGenes,
  loadDownsampledGeneExpression,
};
