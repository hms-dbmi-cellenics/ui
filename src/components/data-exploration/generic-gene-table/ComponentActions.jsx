import React from 'react';
import PropTypes from 'prop-types';
import {
  Menu, Dropdown, Button,
} from 'antd';
import { useSelector, useDispatch } from 'react-redux';

import _ from 'lodash';
import { PlusOutlined, RedoOutlined, MinusOutlined } from '@ant-design/icons';
import { loadGeneExpression } from 'redux/actions/genes';

const geneOperations = {
  ADD: 'add',
  REMOVE: 'remove',
  OVERWRITE: 'overwrite',
};

const ComponentActions = (props) => {
  const { experimentId, name, componentType } = props;
  const dispatch = useDispatch();
  const selectedGenes = useSelector((state) => state.genes.selected);
  const displayedGenes = useSelector((state) => state.genes.expression?.views[componentType]?.data);

  const performGeneOperation = (genesOperation) => {
    let newGenes = _.cloneDeep(selectedGenes);

    if (genesOperation === geneOperations.ADD && displayedGenes) {
      newGenes = Array.from(new Set(displayedGenes.concat(selectedGenes)));
    }
    if (genesOperation === geneOperations.REMOVE && displayedGenes) {
      newGenes = displayedGenes.filter((gene) => !selectedGenes.includes(gene));
    }

    dispatch(loadGeneExpression(experimentId, newGenes, componentType));
  };

  const menu = (
    <Menu size='small'>
      <Menu.Item key='0' icon={<PlusOutlined />} onClick={() => performGeneOperation(geneOperations.ADD)}>
        Add
      </Menu.Item>
      <Menu.Item key='1' icon={<MinusOutlined />} onClick={() => performGeneOperation(geneOperations.REMOVE)}>
        Remove
      </Menu.Item>
      <Menu.Item key='2' icon={<RedoOutlined />} onClick={() => performGeneOperation(geneOperations.OVERWRITE)}>
        Overwrite
      </Menu.Item>
    </Menu>
  );

  if (selectedGenes.length === 0) {
    return (<></>);
  }

  return (
    <Dropdown arrow type='link' size='small' overlay={menu} trigger={['click']}>
      <Button type='link' size='small'>
        {name}
      </Button>
    </Dropdown>
  );
};

ComponentActions.defaultProps = {
};

ComponentActions.propTypes = {
  experimentId: PropTypes.string.isRequired,
  componentType: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
};

export default ComponentActions;
