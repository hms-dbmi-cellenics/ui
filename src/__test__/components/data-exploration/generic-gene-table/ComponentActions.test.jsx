import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { Dropdown } from 'antd';
import waitForActions from 'redux-mock-store-await-actions';
import thunk from 'redux-thunk';
import preloadAll from 'jest-next-dynamic';
import configureMockStore from 'redux-mock-store';
import { GENES_EXPRESSION_LOADING, GENES_EXPRESSION_LOADED } from 'redux/actionTypes/genes';
import { fetchWork } from 'utils/work/fetchWork';
import ComponentActions from 'components/data-exploration/generic-gene-table/ComponentActions';

jest.mock('utils/getTimeoutForWorkerTask', () => ({
  __esModule: true, // this property makes it work
  default: () => 60,
}));

jest.mock('utils/work/fetchWork', () => ({
  fetchWork: jest.fn(() => new Promise((resolve) => resolve({
    A: {
      min: 0,
      max: 1.6,
      expression: [0, 0.4, 0.5, 1.6],
    },
    B: {
      min: 0,
      max: 1.6,
      expression: [0, 0.4, 0.5, 1.6],
    },
    C: {
      min: 0,
      max: 1.6,
      expression: [0, 0.4, 0.5, 1.6],
    },
  }))),
}));

const mockStore = configureMockStore([thunk]);

let component;
const experimentId = '1234';
const componentType = 'asd';

const backendStatus = {
  [experimentId]: {
    status: {
      pipeline: {
        status: 'SUCCEEDED',
        startDate: '2021-01-01T01:01:01.000Z',
      },
    },
  },
};

const initialState = {
  genes: {
    expression: {
      loading: [],
      data: {
        D: {
          min: 0,
          max: 1.6,
          expression: [0, 0.4, 0.5, 1.6],
        },
        E: {
          min: 0,
          max: 1.6,
          expression: [0, 0.4, 0.5, 1.6],
        },
      },
      views: {
        [componentType]: {
          data: ['D', 'E'],
          fetching: false,
          error: false,
        },
      },
    },
    selected: ['A'],
  },
  backendStatus,
};

describe('ComponentActions', () => {
  beforeAll(async () => {
    await preloadAll();
  });

  afterEach(() => {
    component.unmount();
    jest.clearAllMocks();
  });

  it('Renders correctly when there are selected genes', () => {
    const store = mockStore(initialState);

    component = mount(
      <Provider store={store}>
        <ComponentActions name='fakeName' experimentId={experimentId} componentType={componentType} />
      </Provider>,
    );

    const dropdown = component.find(Dropdown);
    expect(dropdown.length).toEqual(1);
    expect(dropdown.props().overlay.type.name).toEqual('Menu');
    expect(dropdown.props().overlay.props.children.length).toEqual(3);
  });

  it('Renders correctly when there are no selected genes', () => {
    const store = mockStore({
      ...initialState,
      genes: {
        ...initialState.genes,
        selected: [],
      },
    });

    component = mount(
      <Provider store={store}>
        <ComponentActions name='fakeName' experimentId={experimentId} componentType={componentType} />
      </Provider>,
    );

    expect(component.find(Dropdown).length).toEqual(0);
  });

  it('Dispatches loadGeneExpression action with the right list of genes when Add is clicked', async () => {
    const store = mockStore({
      ...initialState,
      genes: {
        ...initialState.genes,
        selected: ['A', 'B', 'C'],
      },
    });

    component = mount(
      <Provider store={store}>
        <ComponentActions name='fakeName' experimentId={experimentId} componentType={componentType} />
      </Provider>,
    );

    const menuButtons = component.find(Dropdown).props().overlay;
    menuButtons.props.children[0].props.onClick();

    // Wait for side-effect to propagate (properties loading and loaded).
    await waitForActions(store, [GENES_EXPRESSION_LOADING, GENES_EXPRESSION_LOADED]);

    expect(fetchWork).toHaveBeenCalledWith(
      experimentId,
      {
        name: 'GeneExpression',
        genes: ['A', 'B', 'C'],
      },
      store.getState,
      { timeout: 60 },
    );

    expect(store.getActions().length).toEqual(2);

    expect(store.getActions()[0]).toMatchSnapshot();

    expect(store.getActions()[1]).toMatchSnapshot();
  });

  it('Dispatches loadGeneExpression action with the right list of genes when Remove is clicked', async () => {
    const store = mockStore({
      ...initialState,
      genes: {
        ...initialState.genes,
        selected: ['D'],
      },
    });

    component = mount(
      <Provider store={store}>
        <ComponentActions name='fakeName' experimentId={experimentId} componentType={componentType} />
      </Provider>,
    );

    const menuButtons = component.find(Dropdown).props().overlay;
    menuButtons.props.children[1].props.onClick();

    // Wait for side-effect to propagate (properties loading and loaded).
    await waitForActions(store, [GENES_EXPRESSION_LOADING, GENES_EXPRESSION_LOADED]);

    expect(fetchWork).toHaveBeenCalledTimes(0);

    expect(store.getActions().length).toEqual(2);

    expect(store.getActions()[0]).toMatchSnapshot();

    expect(store.getActions()[1]).toMatchSnapshot();
  });

  it('Dispatches loadGeneExpression action with the right list of genes when Overwrite is clicked', async () => {
    const store = mockStore({
      ...initialState,
      genes: {
        ...initialState.genes,
        selected: ['D'],
      },
    });

    component = mount(
      <Provider store={store}>
        <ComponentActions name='fakeName' experimentId={experimentId} componentType={componentType} />
      </Provider>,
    );

    const menuButtons = component.find(Dropdown).props().overlay;
    menuButtons.props.children[2].props.onClick();

    // Wait for side-effect to propagate (properties loading and loaded).
    await waitForActions(store, [GENES_EXPRESSION_LOADING, GENES_EXPRESSION_LOADED]);

    expect(fetchWork).toHaveBeenCalledTimes(0);

    expect(store.getActions().length).toEqual(2);
    expect(store.getActions()[0]).toMatchSnapshot();
    expect(store.getActions()[1]).toMatchSnapshot();
  });
});
