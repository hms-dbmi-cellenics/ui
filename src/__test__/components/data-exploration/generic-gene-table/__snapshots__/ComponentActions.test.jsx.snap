// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentActions Dispatches loadGeneExpression action with the right list of genes when Add is clicked 1`] = `
{
  "payload": {
    "componentUuid": "asd",
    "experimentId": "1234",
    "genes": [
      "Gzma",
      "Lyz2",
      "GeneA",
      "GeneB",
      "GeneC",
    ],
  },
  "type": "genes/expressionLoading",
}
`;

exports[`ComponentActions Dispatches loadGeneExpression action with the right list of genes when Add is clicked 2`] = `
{
  "payload": {
    "componentUuid": "asd",
    "genes": [
      "Gzma",
      "Lyz2",
      "GeneA",
      "GeneB",
      "GeneC",
    ],
    "newGenes": {
      "orderedGeneNames": [
        "GeneA",
        "GeneB",
        "GeneC",
      ],
      "rawExpression": {
        "datatype": undefined,
        "index": [
          0,
          0,
          2,
          0,
          2,
        ],
        "mathjs": "SparseMatrix",
        "ptr": [
          0,
          1,
          3,
          5,
        ],
        "size": [
          3,
          3,
        ],
        "values": [
          1,
          2,
          5,
          1,
          6,
        ],
      },
      "stats": {
        "rawMean": [
          0.3,
          2.3,
          1.3,
        ],
        "rawStdev": [
          0.4,
          2,
          1,
        ],
        "truncatedMax": [
          1,
          3,
          5,
        ],
        "truncatedMin": [
          0,
          2,
          1,
        ],
      },
      "truncatedExpression": {
        "datatype": undefined,
        "index": [
          0,
          0,
          2,
          0,
          2,
        ],
        "mathjs": "SparseMatrix",
        "ptr": [
          0,
          1,
          3,
          5,
        ],
        "size": [
          3,
          3,
        ],
        "values": [
          1,
          2,
          3,
          0.5,
          1,
        ],
      },
      "zScore": {
        "datatype": undefined,
        "index": [
          0,
          0,
          2,
          0,
          2,
        ],
        "mathjs": "SparseMatrix",
        "ptr": [
          0,
          1,
          3,
          5,
        ],
        "size": [
          3,
          3,
        ],
        "values": [
          1,
          1,
          3,
          3,
          1,
        ],
      },
    },
  },
  "type": "genes/expressionLoaded",
}
`;

exports[`ComponentActions Dispatches loadGeneExpression action with the right list of genes when Overwrite is clicked 1`] = `
{
  "payload": {
    "componentUuid": "asd",
    "experimentId": "1234",
    "genes": [
      "Gzma",
    ],
  },
  "type": "genes/expressionLoaded",
}
`;

exports[`ComponentActions Dispatches loadGeneExpression action with the right list of genes when Remove is clicked 1`] = `
{
  "payload": {
    "componentUuid": "asd",
    "experimentId": "1234",
    "genes": [
      "Lyz2",
    ],
  },
  "type": "genes/expressionLoaded",
}
`;
